File: 2017/day01/day01.cpp
#include "../String.h"
#include "../Task.h"

class Day01 : public AbstractTask {
 public:
  std::string part_one() override {
    return std::to_string(calculate(input(), 1));
  }

  std::string part_two() override {
    auto data = input();
    size_t steps = data.length() / 2;
    return std::to_string(calculate(data, steps));
  }

 private:
  std::string input() { return read_input("inputs/day01_input.txt"); }

  static int calculate(std::string input, int steps) {
    int sum = 0;
    for (size_t i = 0; i < input.length(); i++) {
      size_t next_idx = (i + steps) % input.length();
      if (input[i] == input[next_idx]) sum += c_to_digit(input[i]);
    }

    return sum;
  }
};

File: 2022/src/main/kotlin/se/augustocesar/day01/Day01.kt
package se.augustocesar.day01

import se.augustocesar.Task

class Day01 : Task() {
    override fun partOne(): String {
        return caloriesPerElf(readInput()).maxOrNull().toString()
    }

    override fun partTwo(): String {
        return caloriesPerElf(readInput())
            .sortedDescending()
            .slice(0..2)
            .sum()
            .toString()
    }

    private fun caloriesPerElf(input: String): List<Int> {
        return input
            .split("\n\n")
            .map { group -> group.split('\n').map { it.toInt() } }
            .map { it.sum() }
    }
}
